Mathlib/Computability/DFA.lean

```diff
@@ -163,7 +182,83 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [Set.mem_singleton_iff] at ha' hc'
   substs ha' hc'
   have h := M.evalFrom_of_pow hb hb'
-  rwa [mem_accepts, evalFrom_of_append, evalFrom_of_append, h, hc]
+  rwa [mem_accepts, eval, evalFrom_of_append, evalFrom_of_append, h, hc]
 #align DFA.pumping_lemma DFA.pumping_lemma
 
+section Maps
+
+variable {α' σ' : Type*} (f : α' → α) (g : σ ≃ σ')
+
+/--
+`M.comap f` pulls back the alphabet of `M` along `f`. In other words, it applies `f` to the input
+before passing it to `M`.
+-/
+@[simps]
+def comap : DFA α' σ where
+  step s a := M.step s (f a)
+  start := M.start
+  accept := M.accept
+
+@[simp]
+theorem comap_id : M.comap id = M := rfl
+
+@[simp]
+theorem evalFrom_comap (s : σ) (x : List α') :
+    (M.comap f).evalFrom s x = M.evalFrom s (x.map f) := by
+  induction x using List.list_reverse_induction with
+  | base => simp
+  | ind x a ih => simp [ih]
+
+@[simp]
+theorem eval_comap (x : List α') : (M.comap f).eval x = M.eval (x.map f) := by simp [eval]
+
+@[simp]
+theorem accepts_comap : (M.comap f).accepts = List.map f ⁻¹' M.accepts := by
+  ext x
+  conv =>
+    rhs
+    rw [Set.mem_preimage, mem_accepts]
+  simp [mem_accepts]
+
+/-- Lifts an equivalence on states to an equivalence on DFAs. -/
+@[simps apply_step apply_start apply_accept]
+def reindex : DFA α σ ≃ DFA α σ' where
+  toFun M := {
+    step := fun s a => g (M.step (g.symm s) a)
+    start := g M.start
+    accept := g.symm ⁻¹' M.accept
+  }
+  invFun M := {
+    step := fun s a => g.symm (M.step (g s) a)
+    start := g.symm M.start
+    accept := g ⁻¹' M.accept
+  }
+  left_inv M := by simp
+  right_inv M := by simp
+
+@[simp]
+theorem reindex_refl : reindex (Equiv.refl σ) M = M := rfl
+
+@[simp]
+theorem symm_reindex : (reindex (α := α) g).symm = reindex g.symm := rfl
+
+@[simp]
+theorem evalFrom_reindex (s : σ') (x : List α) :
+    (reindex g M).evalFrom s x = g (M.evalFrom (g.symm s) x) := by
```

Comment:
Why is `(g.symm s)` better than `(g.invFun s)` in the statement? The proof would stay the same.

Reply:
`.invFun` should never appear in any statement

Reply:
The current statement is interpreted as `(reindex g M).evalFrom s x = g (M.evalFrom (g.symm.toFun s) x)` afaik.

Reply:
I don't want to argue against the current version. Just tell me why we don't go with the more elementary version. I am not asking you to change it. Just please explain it to me.

Reply:
> The current statement is interpreted as `(reindex g M).evalFrom s x = g (M.evalFrom (g.symm.toFun s) x)` afaik.

No that's not right. It's interpreted as `(reindex g M).evalFrom s x = ⇑g (M.evalFrom (⇑g.symm s) x)` where `⇑` is `FunLike.coe`.

`g.invFun` is not used because we've decided to use `⇑g.symm` instead as it gives access to more API. That's it, really.

Reply:
I believe the reason is that there are many more lemmas for `.symm`. The only lemmas I could find for `.invFun` are those that convert it to `.symm`.

Also, because `Equiv.invFun_as_coe` is a simp lemma, its RHS (`⇑e.symm`) is considered the normal form. I'm not 100% sure on this, but it'll probably trigger a linter warning if we use `invFun` instead.